### PR TITLE
FIX: Misinterpretation of voxel ordering in LTAs

### PR DESCRIPTION
**Pull request by oesteban**
The structarray was used directly and the extra axis actually changed the order of operations when the direction cosines were scaled by the voxel sizes.

A new test case has been added for which this error was apparent. This bug caused nipreps/fmriprep#2307, nipreps/fmriprep#2393, and nipreps/fmriprep#2410. nipreps/fmriprep#2444 just works around the problem by using ``lta_convert`` instead of *NiTransforms*. The ``lta_convert`` tool can be now dropped.

Resolves: #125

### CODE REVIEW

File: nitransforms/io/lta.py [hunk 1]
```diff
@@ -39,10 +39,10 @@ def as_affine(self):
...
         sa = self.structarr
-        A = np.hstack((sa["xras"], sa["yras"], sa["zras"])) * sa["voxelsize"]
-        b = sa["cras"] - A.dot(sa["volume"]) / 2
-        affine[:3, :3] = A
-        affine[:3, [3]] = b
+        affine[:3, :3] = (
+            np.hstack((sa["xras"], sa["yras"], sa["zras"])) * sa["voxelsize"].flatten()
+        )
+        affine[:3, [3]] = sa["cras"] - affine[:3, :3] @ sa["volume"] / 2
         return affine
```
Comment (effigies) [inline nitransforms/io/lta.py:46]: I feel like this loses clarity. Right now we're decomposing `Mx = Ax + b`, and showing how A and b interact.

Also, if I'm understanding you correctly, the problem is that we're applying `voxelsize` row-wise when it should be column-wise, which suggests the correct fix is:

```Diff
-        A = np.hstack((sa["xras"], sa["yras"], sa["zras"])) * sa["voxelsize"]
+        A = np.hstack((sa["xras"], sa["yras"], sa["zras"])) @ sa["voxelsize"]
```

The reason that the header fields are shaped as they are is to make the computations straightforward linear algebra. If it's not useful, we can just flatten them into shape `(3,)` instead of `(3, 1)`.

Finally, however we do this, could we add a test with a geometry that shows the old formulation was wrong and the new correct? I get the regression tests are checking this, but something readable, with, e.g. voxel sizes like `(2, 3, 4)` and cosines that are ones/zeros would be nice.

File: nitransforms/tests/test_io.py [hunk 2]
```diff
@@ -37,6 +38,22 @@ def test_VolumeGeometry(tmpdir, get_testdata):
...
+-3.0000   0.0000  -0.0000    91.3027
+-0.0000   2.0575  -2.9111   -25.5251
+ 0.0000   2.1833   2.7433  -105.0820
+ 0.0000   0.0000   0.0000     1.0000"""))
+
+    lta_text = "\n".join(
+        (data_path / "bold-to-t1w.lta").read_text().splitlines()[13:21]
+    )
+    r2r = VG.from_string(lta_text)
+    assert np.allclose(r2r.as_affine(), oracle, rtol=1e-4)
```
Comment (effigies) [inline nitransforms/tests/test_io.py:54]: AFAICT this test doesn't fail before this PR. What exactly is the edge case we're fixing here?

Comment (oesteban) [reply nitransforms/tests/test_io.py:54]: Yes, this failed before this PR.

Comment (oesteban) [reply nitransforms/tests/test_io.py:54]: The test breaks current master without the patch proposed -> https://github.com/poldracklab/nitransforms/compare/fix/vg-lta-test

File: nitransforms/io/lta.py (hunk 1 above)
Comment (oesteban) [reply nitransforms/io/lta.py]:
```Python
A = np.hstack((sa["xras"], sa["yras"], sa["zras"])) * sa["voxelsize"]
```
If the dimensions of `sa["voxelsize"]` are (3, 1) then this is row-wise product. Conversely, if the dimension is (3, ) then this is col-wise product. We want column wise product and before this patch we were doing row wise product.

Dot product doesn't apply here, as it rotates the voxelsize vector through the affine.

**Comment by oesteban** [comment]
Good to go? Do you feel the comments have been successfully addressed?

**Comment by effigies** [comment]
Sorry, just getting back to this this morning. Give me one hour.

File: nitransforms/tests/test_io.py (hunk 2 above)
Comment (effigies) [inline nitransforms/tests/test_io.py:54]: Here's the kind of test I was thinking:

```suggestion
def test_volume_group_voxel_ordering():
    vg = VG.from_string("""\
valid = 1  # volume info valid
filename = no_file
volume = 5 6 7
voxelsize = 2 3 4
xras   = -1 0 0
yras   = 0 0 1
zras   = 0 -1 0
cras   = 0 0 0""")
    aff = vg.as_affine()
    assert np.allclose(vg["voxelsize"], [2, 3, 4])
    assert np.allclose(nb.affines.voxel_sizes(aff), [2, 3, 4])
    assert nb.aff2axcodes(aff) == ("L", "S", "P")
```

The property we're testing is that the zooms are correctly applied, so I'd rather be explicit than have it vague what qualities about the problematic geometry have been fixed.

File: nitransforms/io/lta.py [hunk 3]
```diff
@@ -24,26 +24,23 @@ class VolumeGeometry(StringBasedStruct):
...
-        affine = np.eye(4)
         sa = self.structarr
-        A = np.hstack((sa["xras"], sa["yras"], sa["zras"])) * sa["voxelsize"]
-        b = sa["cras"] - A.dot(sa["volume"]) / 2
-        affine[:3, :3] = A
-        affine[:3, [3]] = b
-        return affine
+        R = np.hstack((sa["xras"], sa["yras"], sa["zras"])) * sa["voxelsize"]
+        offset = sa["cras"] - R @ sa["volume"] / 2
+        return from_matvec(R, offset)
```
Comment (effigies) [inline nitransforms/io/lta.py:43]: We should use a consistent notation here and in L74-80. I prefer A/b because this is a common representation of an augmented matrix (see [Affine transformation - Representation](https://en.wikipedia.org/wiki/Affine_transformation#Representation)).

In any event, I wouldn't use R, because that means rotation, and this matrix has scale and rotation. (Not sure if you can sneak shear into the MGH format in practice.)

```suggestion
        A = np.hstack((sa["xras"], sa["yras"], sa["zras"])) * sa["voxelsize"]
        b = sa["cras"] - A @ sa["volume"] / 2
        return from_matvec(A, b)
```

File: nitransforms/tests/test_io.py (hunk 2 above)
Comment (oesteban) [reply nitransforms/tests/test_io.py:54]: Sure, having an extra test and it being clear can only help.

However, I would keep the test I added for two reasons:
- First, because this is one example we know at some point of history failed.
- Second, because the test I wrote is dual to some test freesurfer folks could add to their repo (or, better said, I bet they already are doing something like that). Because `oracle` comes from executing `mri_info` (I saved the part of actually running it on every test), then we are able to at least ensure that our volume geometry interpretation is the same of freesurfer 6.0.1.

Comment (effigies) [reply nitransforms/tests/test_io.py:54]: Sure, we can keep your existing test.

File: nitransforms/io/lta.py [hunk 4]
```diff
@@ -72,15 +69,15 @@ def from_image(klass, img):
         volgeom = klass()
         sa = volgeom.structarr
         sa["valid"] = 1
-        sa["volume"][:, 0] = img.shape[:3]  # Assumes xyzt-ordered image
-        sa["voxelsize"][:, 0] = voxel_sizes(img.affine)[:3]
+        sa["volume"] = img.shape[:3]  # Assumes xyzt-ordered image
+        sa["voxelsize"] = voxel_sizes(img.affine)[:3]
         A = img.affine[:3, :3]
-        b = img.affine[:3, [3]]
+        b = img.affine[:3, [3]].flatten()
```
Comment (effigies) [inline nitransforms/io/lta.py:75]:
```suggestion
        b = img.affine[:3, 3]
```

**Review by effigies — APPROVED** [review]
Looks good. One tiny suggestion to avoid a call to `flatten()`.
